feat(HR): Auto Attendance

karthikeyan5 opened the pull request:
**Employee Checkin Doctype**
![employee_checkin_doctype](https://user-images.githubusercontent.com/4847317/59259963-543ff880-8c58-11e9-970c-302a9055c7c9.png)

- Employee Checkin can be used to store IN/OUT Logs of employees from biometric/RFID devices.
- added '`attendance_device_id`' in the Employee Doctype
- added a function: `erpnext.hr.doctype.employee_checkin.employee_checkin.add_log_based_on_employee_field`
  - this whitelisted method can be called with '`attendance_device_id`' and timestamp to create an Employee Checkin Log.


**Shift Type**
![shift_type_doctype](https://user-images.githubusercontent.com/4847317/59260362-1becea00-8c59-11e9-8e7d-c2d786035e5b.png)
- Auto Attendance settings have been added to Shift Type. You can make the appropriate settings to enable the system to mark attendance based on the 'Employee Checkin'


**HR Settings**
![hr_settings](https://user-images.githubusercontent.com/4847317/59260614-9584d800-8c59-11e9-82e5-c428c0725104.png)
- HR Settings have been regrouped/rearranged and unused fields have been removed

Docs PR: https://github.com/frappe/erpnext_com/pull/398

karthikeyan5 committed:
feat(HR): Added Employee CheckIn Log

> Employee CheckIn Log can be used to store IN/OUT Logs of employees from biometric/RFID devices.
>added biometric_id in the Employee Doctype

karthikeyan5 committed:
fix(HR): changes as per PR review

> renamed 'Employee Checkin Log' DocType to 'Employee_attendance_log' DocType
> renamed biometric_id to biometric_rf_if

karthikeyan5 commented on erpnext/hr/doctype/employee_checkin_log/employee_checkin_log.json:
```diff
@@ -0,0 +1,126 @@
+{
```
done...

Olawale1 commented:
This is a welcome development!

karthikeyan5 committed:
fix(HR): validations and bug fixes for Attendance Log

rmehta commented on erpnext/hr/doctype/employee_attendance_log/employee_attendance_log.py:
```diff
@@ -0,0 +1,45 @@
+
+
+@frappe.whitelist()
+def add_log_based_on_biometric_rf_id(biometric_rf_id, timestamp, device_id=None, log_type=None):
```
fieldname should be `attendance_device_id` not `biometric_rf_id`

It may not be biometric, could be a card scan too!

karthikeyan5 replied:
Yes, It was named `'Biometric/RF ID'` in the UI. I have now changed it to `attendance_device_id`

karthikeyan5 committed:
feat(HR): Auto Attendance

>Marking attendance based on Employee Attendance Log

karthikeyan5 committed:
fix(auto attendance): changes as requested in review
> removed unused field from shift assignment
> renamed 'biometric_rf_id' to 'attendance_device_id'
> added more test cases
> few other minor changes after demo

karthikeyan5 committed:
refactor(HR): Auto Attendance

> moved all auto attendance settings from HR Settings to shift
> added shift in attendance and Employee Attendance Log
> reordered and cleaned up fields in HR Settings and Employee DocType

karthikeyan5 committed:
rename(HR): Employee Attendance Log DocType to Employee Checkin

karthikeyan5 committed:
fix(hr): Auto Attendance

> fixing minor bug in shift calculation
> adding 'Mark Auto Attendance' button in shift and related fixes
> changed employee checkin naming series
> added aditional validation for employee checkin
> added/updated dashboard for Attendance, Employee, Shift Type

karthikeyan5 commented:
Please do not squash...

rmehta commented:
can you fix the tests?

> On 12-Jun-2019, at 12:12 PM, Karthikeyan S <notifications@github.com> wrote:
> 
> Please do not squash...
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/frappe/erpnext/pull/17517?email_source=notifications&email_token=AABCGLFHI57SCYMOTYDPZO3P2CLEPA5CNFSM4HLO2QD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXPNEGA#issuecomment-501142040>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AABCGLFL6VQLK3NM7OUWBXDP2CLEPANCNFSM4HLO2QDQ>.
>

karthikeyan5 commented:
> can you fix the tests?

an unrelated test was failing, and this has been solved by @deepeshgarg007 and merged in Develop. a rebase with develop has now fixed the tests in this PR.

khadijazaki commented:
Which bio metric device can be used if such code is developed?

khadijazaki commented:
Which bio metric device can be used if such code is developed?

flexy2ky commented:
This is absolutely beautiful!!!

karthikeyan5 committed:
fix(HR): fixing bug introduced during merge conflict resolution.

> while merging with https://github.com/frappe/erpnext/pull/17837

karthikeyan5 commented:
> Which bio metric device can be used if such code is developed?

@khadijazaki you can use any device, all ERPNext needs are the biometric_id and timestamp, that you can pass to the API Method '`erpnext.hr.doctype.employee_checkin.employee_checkin.add_log_based_on_employee_field`'... 

Currently, there is a [python script](https://github.com/karthikeyan5/push-biometric-erpnext)(which can do the above) that can be installed in any computer on your local network which has access to the IP Address of your biometric device. This script currently supports most ZKTeco and ESSL Biometric devices...(i.e. any device which uses the ZKProtocol to communicate via an IP Network connection)

Also, an application that can be downloaded, installed and set up biometric devices with no development expertise is in the works. Keep an eye on the [forum](https://discuss.erpnext.com) to know when that is complete.

Olawale1 commented:
Hi @karthikeyan5

Thanks again for the great work. What about Log Type (i.e IN/OUT). Doesn't that need to be included in the api call as well?

Also, any update on the application for users with no development expertise?

Thanks
